### PR TITLE
refactor: extract deploy command logic into separate functions

### PR DESCRIPTION
**Pull request by erezrokah**
**- Summary**

The PR refactors the `deploy` command code by extracting most of the logic into functions.
Some logic mutates objects so I didn't extract it yet since I wanted the functions to be pure.

**- Test plan**

Tested manually the different deploy flags, and actually exposed old bugs in the process (which I fixed).

**- Description for the changelog**

Should not change behaviour

**- A picture of a cute animal (not mandatory but encouraged)**

🐼 

Related to https://github.com/netlify/cli/issues/1236 and https://github.com/netlify/cli/issues/1227 - I initially started looking into those issues and found the code very hard to follow.

### CODE REVIEW

File: src/commands/deploy.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ const path = require('path')
 const chalk = require('chalk')
 const { flags } = require('@oclif/command')
 const get = require('lodash.get')
-const fs = require('fs')
+const fs = require('fs-extra')
```
Comment (erezrokah) [inline src/commands/deploy.js:7]: So we can switch to `async` interfaces

File: src/commands/deploy.js [hunk 2]
```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
 const LinkCommand = require('./link')
 const { NETLIFYDEV, NETLIFYDEVLOG, NETLIFYDEVERR } = require('../utils/logo')
 
+const triggerDeploy = async ({ api, siteId, siteData, log, error }) => {
```
Comment (erezrokah) [inline src/commands/deploy.js:19]: Extracted from https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L89

File: src/commands/deploy.js [hunk 3]
```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
...
+    if (err.status === 404) {
+      error('Site not found. Please rerun "netlify link" and make sure that your site has CI configured.')
+    } else {
+      error(err.message)
+    }
+  }
+}
+
+const getDeployFolder = async ({ flags, config, site, siteData, log }) => {
+  let deployFolder
```
Comment (erezrokah) [inline src/commands/deploy.js:35]: Extracted from https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L105 and https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L125

File: src/commands/deploy.js [hunk 4]
```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
...
+        filter: input => path.resolve(process.cwd(), input),
+      },
+    ])
+    deployFolder = promptPath
+  }
+
+  return deployFolder
+}
+
+const validateDeployFolder = async ({ deployFolder, error }) => {
```
Comment (erezrokah) [inline src/commands/deploy.js:62]: Old version was:
https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L463

Which was broken in multiple places:
1. Using `e.status` instead of `e.code` (fixed in the new version)
https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L468
2. Referencing `stat.isDirectory` instead of `stat.isDirectory()` https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L480

File: src/commands/deploy.js [hunk 5]
```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
...
+    throw e
+  }
+
+  if (!stat.isDirectory()) {
+    return error('Deploy target must be a path to a directory')
+  }
+  return stat
+}
+
+const getFunctionsFolder = ({ flags, config, site, siteData }) => {
```
Comment (erezrokah) [inline src/commands/deploy.js:84]: Extracted from https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L114

File: src/commands/deploy.js [hunk 6]
```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
...
+    functionsFolder = path.resolve(process.cwd(), flags['functions'])
+  } else if (funcConfig) {
+    functionsFolder = path.resolve(site.root, funcConfig)
+  } else if (get(siteData, 'build_settings.functions_dir')) {
+    functionsFolder = path.resolve(site.root, get(siteData, 'build_settings.functions_dir'))
+  }
+  return functionsFolder
+}
+
+const validateFunctionsFolder = async ({ functionsFolder, log, error }) => {
```
Comment (erezrokah) [inline src/commands/deploy.js:98]: Extracted from https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L114 and fixed to reference `stat.isDirectory()`

File: src/commands/deploy.js [hunk 7]
```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
...
+    error('Functions folder must be a path to a directory')
+  }
+
+  return stat
+}
+
+const validateFolders = async ({ deployFolder, functionsFolder, error, log }) => {
+  const deployFolderStat = await validateDeployFolder({ deployFolder, error })
+  const functionsFolderStat = await validateFunctionsFolder({ functionsFolder, error, log })
+  return { deployFolderStat, functionsFolderStat }
```
Comment (erezrokah) [inline src/commands/deploy.js:128]: We need the return value of `validateFunctionsFolder`. If it's undefined (since the folder doesn't exist) we don't pass `functionsFolder` to the `js-client`.
This was done here:
https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L167

File: src/commands/deploy.js [hunk 8]
```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
...
+}
+
+const runDeploy = async ({
```
Comment (erezrokah) [inline src/commands/deploy.js:131]: Extracted from here:
https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L181

File: src/commands/deploy.js [hunk 9]
```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
...
+        return
+      }
+    }
+  }
+
+  const siteUrl = results.deploy.ssl_url || results.deploy.url
+  const deployUrl = get(results, 'deploy.deploy_ssl_url') || get(results, 'deploy.deploy_url')
+  const logsUrl = `${get(results, 'deploy.admin_url')}/deploys/${get(results, 'deploy.id')}`
+
+  return {
```
Comment (erezrokah) [inline src/commands/deploy.js:209]: This information was used here https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L241 and here https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L263

File: src/commands/deploy.js [hunk 10]
```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
...
+    siteId: results.deploy.site_id,
+    siteName: results.deploy.name,
+    deployId: results.deployId,
+    siteUrl,
+    deployUrl,
+    logsUrl,
+  }
+}
+
+const printResults = ({ flags, results, deployToProduction, log, logJson, exit }) => {
```
Comment (erezrokah) [inline src/commands/deploy.js:220]: Extracted from https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L245

File: src/commands/deploy.js [hunk 11]
```diff
@@ -78,216 +321,59 @@ class DeployCommand extends Command {
...
-        fnDir: functionsFolder,
-        statusCb: flags.json || flags.silent ? () => {} : deployProgressCb(),
-        draft: !deployToProduction && !alias,
-        message: flags.message,
-        deployTimeout: flags.timeout * 1000 || 1.2e6,
-        syncFileLimit: 100,
-        branch: alias,
+    log(
+      prettyjson.render({
+        'Deploy path': deployFolder,
```
Comment (erezrokah) [inline src/commands/deploy.js:341]: We don't need the conditions here:
https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L145
and here https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L150 and `prettyjson` doesn't print `undefined` values

File: src/commands/deploy.js (hunk 1 above)
Comment (ehmicky) [reply src/commands/deploy.js:7]: Could we use `utils.promisify` instead?
`fs-extra` includes `graceful-fs` which monkey patches global variables.

File: src/commands/deploy.js [hunk 12]
```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
...
     const alias = flags.alias || flags.branch
 
     if (flags.branch) {
-      this.warn('--branch flag has been renamed to --alias and will be removed in future versions')
+      warn('--branch flag has been renamed to --alias and will be removed in future versions')
     }
 
     const deployToProduction = flags.prod
-    if (deployToProduction && alias) {
-      this.error(`--prod and --alias flags cannot be used at the same time`)
```
Comment (erezrokah) [inline src/commands/deploy.js:31]: Replaced this with https://github.com/netlify/cli/pull/1240/files#diff-e9001bf37e31e33db8a717da3ac59a69R474

File: src/commands/deploy.js (hunk 1 above)
Comment (erezrokah) [reply src/commands/deploy.js:7]: Sure, I can also open an issue to cleanup `fs-extra` from the codebase

Comment (erezrokah) [reply src/commands/deploy.js]: See ef9a68ae34cf4357d2bff859671fd22db3c7fdc1 and See https://github.com/netlify/cli/issues/1241

Comment (ehmicky) [reply src/commands/deploy.js]: Oh I did not know we used it elsewhere :)

File: src/commands/deploy.js (hunk 4 above)
Comment (ehmicky) [reply src/commands/deploy.js:65]: :man_facepalming:

File: src/commands/deploy.js [hunk 13]
```diff
@@ -16,20 +17,265 @@ const SitesCreateCommand = require('./sites/create')
...
+
+  if (stat && !stat.isDirectory()) {
+    error('Functions folder must be a path to a directory')
+  }
+
+  return stat
+}
+
+const validateFolders = async ({ deployFolder, functionsFolder, error, log }) => {
+  const deployFolderStat = await validateDeployFolder({ deployFolder, error })
```
Comment (ehmicky) [inline src/commands/deploy.js:129]: Can we run those two `async` functions in parallel (`Promise.all()`)?

File: src/commands/deploy.js [hunk 14]
```diff
@@ -16,20 +17,265 @@ const SitesCreateCommand = require('./sites/create')
...
+      log('Deploying to draft URL...')
+    }
+
+    results = await api.deploy(siteId, deployFolder, {
+      configPath,
+      fnDir: functionsFolder,
+      statusCb: flags.json || flags.silent ? () => {} : deployProgressCb(),
+      draft: !deployToProduction && !alias,
+      message: flags.message,
+      deployTimeout: flags.timeout * 1000 || 1.2e6,
```
Comment (ehmicky) [inline src/commands/deploy.js:177]: Should we extract that `1.2e6` into a top-level constant variable?

**Review by ehmicky — APPROVED** [review]
Thanks @erezrokah, this is much cleaner! :tada:

File: src/commands/deploy.js (hunk 13 above)
Comment (erezrokah) [reply src/commands/deploy.js:129]: `validateFunctionsFolder` prints a warning message when the folder doesn't exist, while `validateDeployFolder` exists.
Didn't want to run into a case where we print a warning then exit with another error.

File: src/commands/deploy.js (hunk 14 above)
Comment (erezrokah) [reply src/commands/deploy.js:177]: Sure going to update it

File: src/commands/deploy.js (hunk 13 above)
Comment (ehmicky) [reply src/commands/deploy.js:129]: Makes sense :+1: 
This is micro-optimization anyway since those I/O calls are quite fast.

File: src/commands/deploy.js (hunk 14 above)
Comment (erezrokah) [reply src/commands/deploy.js]: Done in https://github.com/netlify/cli/pull/1240/commits/f00db3f9b558720907882f3fe2df9e41e276ea9d#diff-e9001bf37e31e33db8a717da3ac59a69R179
